internal/codeintel/bundles/persistence/cache/reader_cache.go

```diff
@@ -0,0 +1,248 @@
+	entry.wg.Add(1)                         // Mark as in-use
+	element := c.evictList.PushFront(entry) // Update recency data
+	c.entries[key] = element
+
+	// Perform the open procedure in a goroutine. The close of the init channel will signal
+	// all consumers of the cache that the entry's underlying reader is now ready for use.
+	// This allows us to have several consumers concurrently waiting on the same cache entry
+	// to initialize.
+	//
+	// Previous constructions of this cache did not guarantee this property:
+	//   (1) holding a lock during initialization stops independent readers for initializing
+	//       concurrently, which can severely decrease request throughput
+	//   (2) not synchronizing during initialization can allow the same reader to be
+	//       initialized multiple times as there is no "pending" entry in the cache for the
+	//       second request to the same key to wait on.
+	go func() {
+		// Mark as initialized
+		defer close(entry.init)
+
+		entry.reader, entry.err = c.opener(key)
+		if entry.err != nil {
+			// Do not hold on to cache entries that failed to initialize. In the case of a
+			// transient error (e.g., a missing bundle file that is later uploaded), we do
+			// not want a poison cache value occupying that key until a process restart.
+			c.remove(entry)
+		}
+	}()
+
+	// We just created a new entry so we may now be over capcity. Remove as many entries
+	// as necessary to get us back under our maximum size. This may call end up calling
+	// remove on an entry that is already draining, but is safe to do as that function
+	// is idempotent and is guaranteed to remove the underlying entry at some point in
+	// the future.
+	c.evict(len(c.entries) - c.size)
+
+	return entry
+}
+
+// evict attempts to remove n elements from the back of the list.  This function assumes that the cache's
+// mutex is held by the caller.
+func (c *Cache) evict(n int) {
+	for element := c.evictList.Back(); element != nil; element = element.Prev() {
+		if n <= 0 {
+			return
+		}
+
+		n--
+		c.remove(element.Value.(*cacheEntry))
+	}
+}
+
+// remove marks the entry as draining so that future readers will not attempt to hold it,
+// then waits for all current readers to drain. The entry is then removed from the cache
+// mapping. This function is safe to call on the same entry multiple times.
+func (c *Cache) remove(entry *cacheEntry) {
+	entry.once.Do(func() {
+		go func() {
+			entry.close()
+
+			c.m.Lock()
+			defer c.m.Unlock()
+
+			element := c.entries[entry.key]
+			c.evictList.Remove(element)
+			delete(c.entries, entry.key)
+		}()
+	})
+}
+
+// cacheEntry wraps a reader which may still be initializing.
+type cacheEntry struct {
+	key      string
+	reader   persistence.Reader
+	err      error
+	init     chan struct{}  // marks init completion
+	draining chan struct{}  // marks intent to close
+	wg       sync.WaitGroup // concurrent use ref count
+	once     sync.Once      // guards close function
```

Comment:
nit: give them better names, e.g.
```suggestion
	refCount      sync.WaitGroup // concurrent use ref count
	closeOnce     sync.Once      // guards close function
```